From f17e052aba4435e607f5b5d06e7b97a412dda378 Mon Sep 17 00:00:00 2001
From: Vinzenz Feenstra <vfeenstr@redhat.com>
Date: Thu, 10 Apr 2014 12:16:56 +0200
Subject: [PATCH 05/14] Added support for sudo

This patch introduces configure controlled choice of the elevation mechanism
used on the system.

configure has now a new option --with-sudohelper=[usermode|sudo]

Values:
 - usermode: elevation via consolehelper and pam
 - sudo: elevation via sudo

Change-Id: I202c1bbea8c95f5ad12edb4959a7d614c89ea246
Bug-Url: https://bugzilla.redhat.com/1043471
Signed-off-by: Vinzenz Feenstra <vfeenstr@redhat.com>
---
 Makefile.am                                   |    4 +---
 configure.ac                                  |   22 +++++++++++++++++
 ovirt-guest-agent/consoleapps/Makefile.am     |   17 +++++++++----
 ovirt-guest-agent/pam/Makefile.am             |   18 +++++++++-----
 scripts/Makefile.am                           |   13 ++++++++++
 scripts/sudoers.ovirt-guest-agent             |    8 +++++++
 scripts/wrappers/Makefile.am                  |   32 +++++++++++++++++++++++++
 scripts/wrappers/ovirt-hibernate-wrapper.sh   |   22 +++++++++++++++++
 scripts/wrappers/ovirt-locksession-wrapper.sh |   22 +++++++++++++++++
 scripts/wrappers/ovirt-shutdown-wrapper.sh    |   22 +++++++++++++++++
 scripts/wrappers/ovirt-sudo-wrapper.sh        |   22 +++++++++++++++++
 11 files changed, 188 insertions(+), 14 deletions(-)
 create mode 100644 scripts/Makefile.am
 create mode 100644 scripts/sudoers.ovirt-guest-agent
 create mode 100644 scripts/wrappers/Makefile.am
 create mode 100755 scripts/wrappers/ovirt-hibernate-wrapper.sh
 create mode 100755 scripts/wrappers/ovirt-locksession-wrapper.sh
 create mode 100755 scripts/wrappers/ovirt-shutdown-wrapper.sh
 create mode 100755 scripts/wrappers/ovirt-sudo-wrapper.sh

diff --git a/Makefile.am b/Makefile.am
index ff73c6a..744a403 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -5,6 +5,7 @@ SUBDIRS =                \
     GinaSSO              \
     tests                \
     windows-credprov     \
+    scripts              \
     $(NULL)
 
 if BUILD_SSO_MODULES
@@ -75,7 +76,4 @@ endif
 
 install-exec-hook:
 	$(MKDIR_P) $(DESTDIR)/$(pkgdatadir)
-	$(LN_S) -f $(bindir)/consolehelper $(DESTDIR)/$(pkgdatadir)/ovirt-locksession
-	$(LN_S) -f $(bindir)/consolehelper $(DESTDIR)/$(pkgdatadir)/ovirt-shutdown
-	$(LN_S) -f $(bindir)/consolehelper $(DESTDIR)/$(pkgdatadir)/ovirt-hibernate
 	$(INSTALL) -d $(DESTDIR)/$(localstatedir)/log/ovirt-guest-agent
diff --git a/configure.ac b/configure.ac
index 75183fa..1a3a6f3 100644
--- a/configure.ac
+++ b/configure.ac
@@ -41,6 +41,26 @@ AC_ARG_WITH(kdm,
                            [Whether or not to build the single sign on KDM extension, default=yes]),
             kdm=${withval}, kdm=$sso)
 AC_SUBST(kdm)
+
+AC_ARG_WITH(sudohelper,
+            AS_HELP_STRING([--with-sudohelper],
+                           [[Choose the elevation mechanism to be installed. Available options are usermode or sudo], default=usermode]),
+            sudohelper=${withval}, sudohelper=usermode)
+AC_SUBST(sudohelper)
+
+case $sudohelper in
+    usermode)
+        ;;
+    sudo)
+        ;;
+    *)
+        AC_MSG_ERROR([sudohelper must be one of usermode or sudo])
+        ;;
+esac
+
+AM_CONDITIONAL(INSTALL_USERMODE_SCRIPTS, [test -n "$sudohelper" -a "$sudohelper" == "usermode"])
+AM_CONDITIONAL(INSTALL_SUDO_SCRIPTS, [test -n "$sudohelper" -a "$sudohelper" == "sudo"])
+
 AM_CONDITIONAL(BUILD_GDM_MODULE, [test -n "$gdm" -a "x$gdm" != xno ])
 AM_CONDITIONAL(BUILD_KDM_MODULE, [test -n "$kdm" -a "x$kdm" != xno ])
 AM_CONDITIONAL(BUILD_SSO_MODULES, [test -n "$sso" -a "x$sso" != xno ])
@@ -213,6 +233,8 @@ AC_CONFIG_FILES([
     ovirt-guest-agent/ovirt-guest-agent
     ovirt-guest-agent/consoleapps/Makefile
     ovirt-guest-agent/pam/Makefile
+    scripts/Makefile
+    scripts/wrappers/Makefile
     tests/Makefile
     windows-credprov/Makefile
 ])
diff --git a/ovirt-guest-agent/consoleapps/Makefile.am b/ovirt-guest-agent/consoleapps/Makefile.am
index 200c2dc..097b783 100644
--- a/ovirt-guest-agent/consoleapps/Makefile.am
+++ b/ovirt-guest-agent/consoleapps/Makefile.am
@@ -1,11 +1,18 @@
 
+if INSTALL_USERMODE_SCRIPTS
 consoleappsdir = $(sysconfdir)/security/console.apps
-consoleapps_DATA = \
-    ovirt-shutdown \
-    ovirt-locksession \
-    ovirt-hibernate
+consoleapps_DATA =      \
+    ovirt-shutdown      \
+    ovirt-locksession   \
+    ovirt-hibernate     \
+    $(NULL)
+endif
 
-EXTRA_DIST = $(consoleapps_DATA)
+EXTRA_DIST =            \
+    ovirt-shutdown      \
+    ovirt-locksession   \
+    ovirt-hibernate     \
+    $(NULL)
 
 CLEANFILES = \
     *~
diff --git a/ovirt-guest-agent/pam/Makefile.am b/ovirt-guest-agent/pam/Makefile.am
index 9c65bde..486de87 100644
--- a/ovirt-guest-agent/pam/Makefile.am
+++ b/ovirt-guest-agent/pam/Makefile.am
@@ -1,11 +1,17 @@
-
+if INSTALL_USERMODE_SCRIPTS
 pamdir = $(PAM_PREFIX)/pam.d
-pam_DATA = \
-    ovirt-shutdown \
-    ovirt-locksession \
-    ovirt-hibernate
+pam_DATA =             \
+    ovirt-shutdown     \
+    ovirt-locksession  \
+    ovirt-hibernate    \
+    $(NULL)
+endif
 
-EXTRA_DIST = $(pam_DATA)
+EXTRA_DIST =           \
+    ovirt-shutdown     \
+    ovirt-locksession  \
+    ovirt-hibernate    \
+    $(NULL)
 
 CLEANFILES = \
     *~
diff --git a/scripts/Makefile.am b/scripts/Makefile.am
new file mode 100644
index 0000000..83b8cb4
--- /dev/null
+++ b/scripts/Makefile.am
@@ -0,0 +1,13 @@
+SUBDIRS =    \
+    wrappers \
+    $(NULL)
+
+EXTRA_DIST=                    \
+    sudoers.ovirt-guest-agent  \
+    $(NULL)
+
+if INSTALL_SUDO_SCRIPTS
+install-data-hook:
+	$(MKDIR_P) $(DESTDIR)/$(sysconfdir)/sudoers.d
+	$(INSTALL) -m 440 sudoers.ovirt-guest-agent $(DESTDIR)/$(sysconfdir)/sudoers.d/50_ovirt-guest-agent
+endif
diff --git a/scripts/sudoers.ovirt-guest-agent b/scripts/sudoers.ovirt-guest-agent
new file mode 100644
index 0000000..5539f00
--- /dev/null
+++ b/scripts/sudoers.ovirt-guest-agent
@@ -0,0 +1,8 @@
+Cmnd_Alias OVIRTAGENT_SCRIPTS =\
+    /usr/share/ovirt-guest-agent/ovirt-hibernate-wrapper.sh *,\
+    /usr/share/ovirt-guest-agent/ovirt-shutdown-wrappers.h *,\
+    /usr/share/ovirt-guest-agent/ovirt-locksession-wrapper.sh
+
+ovirtagent ALL= NOPASSWD: OVIRTAGENT_SCRIPTS
+Defaults:ovirtagent !requiretty
+Defaults:ovirtagent !syslog
diff --git a/scripts/wrappers/Makefile.am b/scripts/wrappers/Makefile.am
new file mode 100644
index 0000000..3d87087
--- /dev/null
+++ b/scripts/wrappers/Makefile.am
@@ -0,0 +1,32 @@
+
+EXTRA_DIST =                         \
+    ovirt-hibernate-wrapper.sh       \
+    ovirt-locksession-wrapper.sh     \
+    ovirt-shutdown-wrapper.sh        \
+    ovirt-sudo-wrapper.sh            \
+    $(NULL)
+
+CLEANFILES = \
+    *~
+
+if INSTALL_SUDO_SCRIPTS
+sudoscriptsdir = $(pkgdatadir)
+sudoscripts_SCRIPTS=                \
+    ovirt-sudo-wrapper.sh           \
+    ovirt-hibernate-wrapper.sh      \
+    ovirt-shutdown-wrapper.sh       \
+    ovirt-locksession-wrapper.sh    \
+    $(NULL)
+
+install-exec-hook:
+	$(LN_S) -f $(pkgdatadir)/ovirt-sudo-wrapper.sh $(DESTDIR)/$(pkgdatadir)/ovirt-hibernate
+	$(LN_S) -f $(pkgdatadir)/ovirt-sudo-wrapper.sh $(DESTDIR)/$(pkgdatadir)/ovirt-locksession
+	$(LN_S) -f $(pkgdatadir)/ovirt-sudo-wrapper.sh $(DESTDIR)/$(pkgdatadir)/ovirt-shutdown
+endif
+
+if INSTALL_USERMODE_SCRIPTS
+install-exec-hook:
+	$(LN_S) -f $(bindir)/consolehelper $(DESTDIR)/$(pkgdatadir)/ovirt-locksession
+	$(LN_S) -f $(bindir)/consolehelper $(DESTDIR)/$(pkgdatadir)/ovirt-shutdown
+	$(LN_S) -f $(bindir)/consolehelper $(DESTDIR)/$(pkgdatadir)/ovirt-hibernate
+endif
diff --git a/scripts/wrappers/ovirt-hibernate-wrapper.sh b/scripts/wrappers/ovirt-hibernate-wrapper.sh
new file mode 100755
index 0000000..0f730b7
--- /dev/null
+++ b/scripts/wrappers/ovirt-hibernate-wrapper.sh
@@ -0,0 +1,22 @@
+#!/bin/sh
+#
+#       Copyright (C) 2014 Vinzenz Feenstra, Red Hat, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Refer to the README and COPYING files for full details of the license.
+#
+
+/usr/share/ovirt-guest-agent/hibernate $@
+
+
diff --git a/scripts/wrappers/ovirt-locksession-wrapper.sh b/scripts/wrappers/ovirt-locksession-wrapper.sh
new file mode 100755
index 0000000..da8114b
--- /dev/null
+++ b/scripts/wrappers/ovirt-locksession-wrapper.sh
@@ -0,0 +1,22 @@
+#!/bin/sh
+#
+#       Copyright (C) 2014 Vinzenz Feenstra, Red Hat, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Refer to the README and COPYING files for full details of the license.
+#
+
+/usr/share/ovirt-guest-agent/LockActiveSession.py $@ &> /dev/null
+
+
diff --git a/scripts/wrappers/ovirt-shutdown-wrapper.sh b/scripts/wrappers/ovirt-shutdown-wrapper.sh
new file mode 100755
index 0000000..39cf113
--- /dev/null
+++ b/scripts/wrappers/ovirt-shutdown-wrapper.sh
@@ -0,0 +1,22 @@
+#!/bin/sh
+#
+#       Copyright (C) 2014 Vinzenz Feenstra, Red Hat, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Refer to the README and COPYING files for full details of the license.
+#
+
+/sbin/shutdown $@
+
+
diff --git a/scripts/wrappers/ovirt-sudo-wrapper.sh b/scripts/wrappers/ovirt-sudo-wrapper.sh
new file mode 100755
index 0000000..1218410
--- /dev/null
+++ b/scripts/wrappers/ovirt-sudo-wrapper.sh
@@ -0,0 +1,22 @@
+#!/bin/sh
+#
+#       Copyright (C) 2014 Vinzenz Feenstra, Red Hat, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Refer to the README and COPYING files for full details of the license.
+#
+
+sudo "/usr/share/ovirt-guest-agent/`basename $0`-wrapper.sh" $@
+
+
-- 
1.7.10.4

